### ISSUE TITLE
Packet timeout DoS concerns

### ISSUE THREAD

**Issue by Thunnini** [cosmos/ibc#354]
In current spec, the ordered channel will be closed if a packet is received after a timeout.
But, I worry the case that the attacker do DoS to prevent a relay from delivering packets.
Let's assume that 20-transfer is executed in an ordered channel to give an example.
Attackers can DoS to the source and destination chain. So, It is difficult for a relay to deliver a packet timely. Packet's timeout height may be calculated from the destination chain's client on the source chain. But, it can be old data because the relayer can't deliver the destination chain's header due to a DoS attack. So, it may be tight for a relayer to deliver a packet before the timeout and it is more difficult to deliver a packet due to DoS attack on destination chain. In this case, the channel will be closed and the assets in destination chains will be locked and can't be recovered even though there are no problems in the source chain's consensus and the destination chain's consensus.
Is there an essential reason for closing a channel when a timeout occurs?

**Comment by cwgoes** [comment]
The specific use-case of token transfers does not require an ordered channel, and in general applications that *do* want recovery from timeouts should not use an ordered channel - the idea of an ordered channel is precisely that you must receive each packet in order, and if you don't you have to abort anyways.

**Comment by Thunnini** [comment]
I think it is reasonable to run the interchain account in the ordered channel to process the requested interchain tx in order.
Interchain can be used for the staking derivatives or a decentralized CDP position (i.e. Kava) and so on...
And, an interchain account can be managed by only the chain that has made it.
The more the chain manages assets in counterparty chain, the more possibility the attackers want to attack. And, if the ordered channel is closed, the chain can't recover the closed channel. And, the assets that manage by interchain account will be locked.

**Comment by cwgoes** [comment]
Yes, but if you miss a transaction (on an interchain account) due to a timeout, you don't necessarily want to process the next transaction (since there might have been a state dependency) - I think you instead want some custom "retry/resume" logic that can try the timed-out transaction again, but I don't know if that should be within the scope of core IBC - it seems like you want to create a new channel & associate the interchain account with it, potentially, or have another unordered channel for control messages.

**Comment by Thunnini** [comment]
I think handling an interchain transaction being reverted due to its internal transaction logic and a packet failing due to timeout as the same thing could be useful in certain use cases.
In the interchain account module, if the delivered interchain transaction failed due to the failure of the transaction logic, it should revert the state transition except for the state transition related to the IBC packet (like next receive sequence). If the packet failed due to a timeout, naturally, there will be no state transitions. So, I think we can handle the both case as same.
Are there any side effects that I could have missed for a packet-timeout happening?
If the need to close the channel when packet-timeout happens is not required feature for the IBC specification, I think it is better to leave the decision to either close the channel or leave the channel as is to the module.

**Comment by cwgoes** [comment]
> In the interchain account module, if the delivered interchain transaction failed due to the failure of the transaction logic, it should revert the state transition except for the state transition related to the IBC packet (like next receive sequence). If the packet failed due to a timeout, naturally, there will be no state transitions. So, I think we can handle the both case as same.

Yes, that is true in this case, but you still want some "exceptional" recovery, right? IBC doesn't auto-retry packets, so if there were a state dependency of the next transactions (from the interchain account) on the results of executing this one, it wouldn't work.

cc @warner Do you have an opinion on this use-case / how would Agoric handle it?

**Comment by dtribble** [comment]
A few comments on ways to address this:

In a hostile environment, any provision for retry, etc. is basically "make the timeout longer". How long are you willing to wait for the DoS to subside before you decide you need to give up?  minutes? hours? days?  If you are willing to wait days for it all to get sorted out, then make that the channel timeout.

Regardless, the structure of resources transfer MUST have recovery semantics and behavior for if packets timeout and/or the channel is closed. In the resource transfer case, why are they locked? Or on timeout, they should revert and unlock the resources. the semantics of timeout are such that this can be done reliably by the sender (because they saw a consensus proof of the counterparty after the timeout that does not include accepting the specific message).

Note that packets sent after a channel closes are treated just like timeouts, so there aren't many extra cases to sort out.  Additionally, most of the same issues arise whether it's ordered or unordered; e.g., how long are you willing to wait for a single packet transfer to complete in an unordered channel before you decided it didn't work?

**Comment by cwgoes** [comment]
> Note that packets sent after a channel closes are treated just like timeouts, so there aren't many extra cases to sort out. Additionally, most of the same issues arise whether it's ordered or unordered; e.g., how long are you willing to wait for a single packet transfer to complete in an unordered channel before you decided it didn't work?

The crucial difference - as I understand the likely application usage patterns - is that unordered channels have no *inter-packet state dependencies* - so timeouts can vary between packets according to the specific data being sent & preferences of the sender.

It seems like ordered channels should instead have a timeout set *per-channel*, if they are to be used by multiple users safely (else a user could grief by choosing a low timeout & closing the channel). This is actually not true in the spec at the moment - both ordered & unordered channels have per-packet timeouts chosen by the sender of the packet. Do you think we should change that for ordered channels? I think we probably should.

**Comment by dtribble** [comment]
We discussed the per-packet vs. per-channel timeout in June. One of the observations is that for simplicity and security, we would like to use only one timeout mechanism if that's sufficient.

Many of the arguments for per-channel timeout would apply to unordered channels. At this level, timeouts should largely  be tied to the peculiarities of the infrastructure, and *not* anything application related. That's largely going to be the same for each packet until we have more specialized application packets. So that's a reason to continue to have the implementation use per-packet timeouts, even if the API has a setting for it on the channel. I could easily imagine that API being usable for unordered as well, with the ability to override it on a per-packet basis.  For ordered channels, I could imagine setup packets having a short timeout, and ongoing packets having a longer-term timeout to cope with network vagarities.

All the scenarios I think in terms of either don't have single channels shared between users or don't allow *users* to control the timeout of their app packets (that's controlled in the infrastructure of sending their packets on shared channel. Do you have use case examples for that?

**Comment by cwgoes** [comment]
> All the scenarios I think in terms of either don't have single channels shared between users or don't allow users to control the timeout of their app packets (that's controlled in the infrastructure of sending their packets on shared channel. Do you have use case examples for that?

It's certainly possible to architect things that way - I suppose applications with shared channels could just impose the timeout restrictions themselves, if they allow users (some) control over them, and that's probably fine - one use-case is just token transfers where users might have different preferences for longer/shorter timeouts (since the timeout also controls when they can retrieve escrowed funds if the packet failed to go through).

**Comment by cwgoes** [comment]
@Thunnini Are there any unresolved concerns here that still need to be addressed?

**Comment by Thunnini** [comment]
Can't think of anything more. I think it is good enough for now.

**Comment by cwgoes** [comment]
Alright - thanks for the discussion!

**Comment by gjermundgaraba** [comment]
This issue was mistakenly reopened during some integration work, so closing again now.
